File: api/src/repositories/rate-limits-repository.js
const cosmosClient = require('./cosmos-client');

class RateLimitsRepository {
  async initialize() {
    await cosmosClient.initialize();
    this.container = cosmosClient.getContainer('rateLimits');
  }

  async getTodayRecord() {
    const today = new Date().toISOString().split('T')[0];
    
    try {
      const { resources } = await this.container.items
        .query({
          query: 'SELECT * FROM c WHERE c.date = @date',
          parameters: [{ name: '@date', value: today }]
        })
        .fetchAll();
      
      return resources.length > 0 ? resources[0] : null;
    } catch (error) {
      console.error('Error fetching rate limit record:', error);
      throw error;
    }
  }

  async createOrUpdateRecord(status) {
    const today = new Date().toISOString().split('T')[0];
    const existingRecord = await this.getTodayRecord();

    if (existingRecord) {
      // 既存レコードがある場合
      if (existingRecord.status === 'pending' || existingRecord.status === 'running') {
        // すでに実行中の場合はそのまま返す
        return {
          isAlreadyRunning: true,
          record: existingRecord
        };
      }

      // completed状態の場合はcountを増やして新しいリクエストとして処理
      const updatedRecord = {
        ...existingRecord,
        count: (existingRecord.count || 0) + 1,
        status: status,
        lastRequestedAt: new Date().toISOString()
      };

      const { resource } = await this.container
        .item(existingRecord.id, today)
        .replace(updatedRecord);

      return {
        isAlreadyRunning: false,
        record: resource
      };
    } else {
      // 新規レコードを作成
      const newRecord = {
        id: `rate-limit-${today}`,
        date: today,
        count: 1,
        status: status,
        lastRequestedAt: new Date().toISOString(),
        createdAt: new Date().toISOString()
      };

      const { resource } = await this.container.items.create(newRecord);
      
      return {
        isAlreadyRunning: false,
        record: resource
      };
    }
  }

  async updateStatus(recordId, date, status) {
    try {
      const { resource: existingRecord } = await this.container
        .item(recordId, date)
        .read();

      if (!existingRecord) {
        throw new Error(`Record not found: ${recordId}`);
      }

      const updatedRecord = {
        ...existingRecord,
        status: status,
        updatedAt: new Date().toISOString()
      };

      const { resource } = await this.container
        .item(recordId, date)
        .replace(updatedRecord);

      return resource;
    } catch (error) {
      console.error('Error updating rate limit status:', error);
      throw error;
    }
  }

  async getAllRecords() {
    try {
      await this.initialize();
      const { resources } = await this.container.items
        .readAll()
        .fetchAll();
      
      // 日付でソートして返す
      return resources.sort((a, b) => b.date.localeCompare(a.date));
    } catch (error) {
      console.error('Error fetching all rate limit records:', error);
      throw new Error(`Failed to get rate limit records from Cosmos DB: ${error.message}`);
    }
  }

  async getRecord(date) {
    try {
      await this.initialize();
      const { resources } = await this.container.items
        .query({
          query: 'SELECT * FROM c WHERE c.date = @date',
          parameters: [{ name: '@date', value: date }]
        })
        .fetchAll();
      
      if (resources.length === 0) {
        throw new Error(`Rate limit record for ${date} not found`);
      }
      
      return resources[0];
    } catch (error) {
      if (error.message.includes('not found')) {
        throw error;
      }
      console.error('Error fetching rate limit record:', error);
      throw new Error(`Failed to get rate limit record from Cosmos DB: ${error.message}`);
    }
  }

  async deleteRecordById(id) {
    try {
      await this.initialize();
      
      // IDから日付部分を抽出（パーティションキー用）
      // ID形式: rate-limit-YYYY-MM-DD
      const match = id.match(/^rate-limit-(\d{4}-\d{2}-\d{2})$/);
      if (!match) {
        throw new Error(`Invalid ID format: ${id}`);
      }
      const date = match[1];
      
      // 削除実行（パーティションキーはdate）
      await this.container.item(id, date).delete();
      
      return { 
        success: true, 
        message: `Rate limit record with ID ${id} deleted successfully` 
      };
    } catch (error) {
      if (error.code === 404) {
        console.warn(`Rate limit record not found: ${id}`);
        throw new Error(`Rate limit record with ID ${id} not found`);
      }
      console.error('Failed to delete rate limit record:', error);
      throw new Error(`Failed to delete rate limit record from Cosmos DB: ${error.message}`);
    }
  }
}

const rateLimitsRepository = new RateLimitsRepository();
module.exports = rateLimitsRepository;

File: scraper/src/repositories/rate_limits_repository.py
"""
rate_limitsコンテナへのアクセス管理
"""
import os
from datetime import datetime
from typing import Dict, Optional, Any
from azure.cosmos import CosmosClient, exceptions
from pathlib import Path
from dotenv import load_dotenv

# ルートの.envファイルを読み込み
root_env_path = Path(__file__).parent.parent.parent.parent / '.env'
load_dotenv(root_env_path)


class RateLimitsRepository:
    """rate_limitsコンテナへのアクセスクラス"""
    
    def __init__(self):
        """
        Cosmos DB接続の初期化
        """
        endpoint = os.getenv('COSMOS_ENDPOINT')
        key = os.getenv('COSMOS_KEY')
        database_name = os.getenv('COSMOS_DATABASE', 'studio-reservations')
        
        if not endpoint or not key:
            raise ValueError("Cosmos DB connection settings are missing")
        
        self.client = CosmosClient(endpoint, key)
        self.database = self.client.get_database_client(database_name)
        self.container = self.database.get_container_client('rate_limits')
    
    def get_today_record(self) -> Optional[Dict[str, Any]]:
        """
        本日のレコードを取得
        
        Returns:
            本日のレコード、存在しない場合はNone
        """
        today = datetime.now().strftime('%Y-%m-%d')
        
        try:
            query = "SELECT * FROM c WHERE c.date = @date"
            parameters = [{"name": "@date", "value": today}]
            
            items = list(self.container.query_items(
                query=query,
                parameters=parameters,
                enable_cross_partition_query=True
            ))
            
            return items[0] if items else None
            
        except exceptions.CosmosHttpResponseError as e:
            print(f"Cosmos DB error while fetching rate limit record: {e.message}")
            raise
        except Exception as e:
            print(f"Unexpected error while fetching rate limit record: {e}")
            raise
    
    def create_or_update_record(self, status: str = 'running') -> Dict[str, Any]:
        """
        レコードを作成または更新
        
        Args:
            status: 設定するステータス（'running', 'completed', 'failed'）
        
        Returns:
            dict: {
                'is_already_running': bool,  # すでに実行中の場合True
                'record': dict  # レコード情報
            }
        """
        today = datetime.now().strftime('%Y-%m-%d')
        
        try:
            existing_record = self.get_today_record()
            
            if existing_record:
                # 既存レコードがある場合
                if existing_record.get('status') == 'running':
                    # すでに実行中の場合はそのまま返す
                    return {
                        'is_already_running': True,
                        'record': existing_record
                    }
                
                # completed/failedの場合はcountを増やして新しいリクエストとして処理
                updated_record = {
                    **existing_record,
                    'count': existing_record.get('count', 0) + 1,
                    'status': status,
                    'lastRequestedAt': datetime.now().isoformat() + 'Z',
                    'updatedAt': datetime.now().isoformat() + 'Z'
                }
                
                # partitionKeyを明示的に設定
                self.container.upsert_item(body=updated_record)
                
                return {
                    'is_already_running': False,
                    'record': updated_record
                }
            else:
                # 新規レコードを作成
                new_record = {
                    'id': today,  # idとdateを同じに
                    'date': today,
                    'count': 1,
                    'status': status,
                    'lastRequestedAt': datetime.now().isoformat() + 'Z',
                    'createdAt': datetime.now().isoformat() + 'Z',
                    'updatedAt': datetime.now().isoformat() + 'Z'
                }
                
                self.container.create_item(body=new_record)
                
                return {
                    'is_already_running': False,
                    'record': new_record
                }
                
        except exceptions.CosmosHttpResponseError as e:
            print(f"Cosmos DB error while creating/updating rate limit: {e.message}")
            raise
        except Exception as e:
            print(f"Unexpected error while creating/updating rate limit: {e}")
            raise
    
    def update_status(self, record_id: str, date: str, status: str) -> Dict[str, Any]:
        """
        ステータスを更新
        
        Args:
            record_id: レコードID
            date: 日付（パーティションキー）
            status: 新しいステータス（'completed' or 'failed'）
        
        Returns:
            更新されたレコード
        """
        try:
            # レコードを取得
            existing_record = self.container.read_item(
                item=record_id,
                partition_key=date
            )
            
            if not existing_record:
                raise ValueError(f"Record not found: {record_id}")
            
            # ステータスを更新
            updated_record = {
                **existing_record,
                'status': status,
                'updatedAt': datetime.now().isoformat() + 'Z'
            }
            
            # 更新を実行
            self.container.replace_item(
                item=record_id,
                body=updated_record
            )
            
            return updated_record
            
        except exceptions.CosmosResourceNotFoundError:
            print(f"Rate limit record not found: {record_id}")
            raise ValueError(f"Record not found: {record_id}")
        except exceptions.CosmosHttpResponseError as e:
            print(f"Cosmos DB error while updating status: {e.message}")
            raise
        except Exception as e:
            print(f"Unexpected error while updating status: {e}")
            raise